### ISSUE TITLE
High Availability Isn't Working with BYO

### ISSUE THREAD

**Issue by JohnTheodore** [openshift/openshift-ansible#718]
I am getting "Error: Username and/or password is incorrect" as part of the task "openshift_master_cluster | Authenticate to the cluster"

Though it already failed on the task "Test if cluster is already configured" with the stdout error: "Error: cluster is not currently running on this node"

The logs with vvvvv are here: https://gist.github.com/JohnTheodore/7cc0fb98c5b96cb1d6cc

I can consistently get a single node master setup working. The HA stuff has been troublesome though. I have a hardware load balancer pointing to these three nodes in a round robin fashion. So paas.phx2.mydomain.net points to each of the three master nodes. I use the hardware load balancer with a single node master and that works fine.

Are there any e2e tests around validating if HA deployment works for openshift-ansible? or has anyone else deployed the HA mode without issues?

**Comment by abutcher** [comment]
Hey @JohnTheodore - looks like this is failing because the pcs package was already installed. The cluster password is only set when pcs is installed by ansible - which is something that should definitely be changed.

I'm not sure that you'll want to use this HA method with a hardware load balancer though. The current HA deployment configures a VIP (self assigned) on the acting master node which is the endpoint for the cluster. So, only one master would be active at a time in your lb.

We're working to support HA natively (PR is: https://github.com/openshift/openshift-ansible/pull/608) but there are a few outstanding bits that need updating which are waiting on some upstream changes. If you're feeling froggy, you could give this pull request a try with deployment_type=origin and an inventory file that looks something like this:

```
[OSEv3:children]
masters
nodes
etcd

[OSEv3:vars]                                                                                                                                                                                                    
ansible_ssh_user=root

deployment_type=origin

# Master cluster native ha variables with an optional load balancer                                                                                                                                               
# If no lb group is defined installer assumes that a load balancer has                                                                                                                                            
# been preconfigured. For installation the value of                                                                                                                                                               
# openshift_master_cluster_hostname must resolve to the load balancer                                                                                                                                             
# or to one or all of the masters defined in the inventory if no load                                                                                                                                             
# balancer is present.                                                                                                                                                                                            
openshift_master_cluster_ha=native                                                                                                                                                                               
openshift_master_cluster_hostname=<hostname of vip>
openshift_master_cluster_public_hostname=<hostname of vip>
openshift_master_cluster_vip=<ip of first master to be used for skydns>
openshift_master_cluster_public_vip=<ip of first master to be used for skydnsS>

[masters]
...

[nodes]
...

[etcd]
...
```

This will start the api servers on all masters such that they can be load balanced and start the controllers service on a single master. If the master running the controllers service were to die another would take its place. The pull request will create an haproxy load balancer when an `[lb]` group is defined in the inventory.

**Comment by abutcher** [comment]
Enterprise should work with the PR too since the native changes are in 3.0.2 and obviously the skydns endpoint is the last piece of the puzzle...

**Comment by JohnTheodore** [comment]
@abutcher I did a run on brand new centos 7 nodes with the config setup how you described. It's been hanging on this step for the last 3 hours: 

TASK: [openshift_facts | Gather Cluster facts] ********************************
<b-openshif-r18f10-prod.phx2.mydomain.net> ESTABLISH CONNECTION FOR USER: johntheodore
<b-openshif-r18f10-prod.phx2.mydomain.net> REMOTE_MODULE openshift_facts
<b-openshif-r18g8-prod.phx2.mydomain.net> ESTABLISH CONNECTION FOR USER: johntheodore
<b-openshif-r18g8-prod.phx2.mydomain.net> REMOTE_MODULE openshift_facts
<b-openshif-r18f8-prod.phx2.mydomain.net> ESTABLISH CONNECTION FOR USER: johntheodore
<b-openshif-r18f8-prod.phx2.mydomain.net> REMOTE_MODULE openshift_facts
<b-openshif-r18f10-prod.phx2.mydomain.net> EXEC sshpass -d13 ssh -C -tt -vvv -o ControlMaster=auto -o ControlPersist=60s -o ControlPath="/home/johntheodore/.ansible/cp/ansible-ssh-%h-%p-%r" -o GSSAPIAuthentication=no -o PubkeyAuthentication=no -o ConnectTimeout=10 b-openshif-r18f10-prod.phx2.mydomain.net /bin/sh -c 'mkdir -p $HOME/.ansible/tmp/ansible-tmp-1445471371.06-1971578243438 && chmod a+rx $HOME/.ansible/tmp/ansible-tmp-1445471371.06-1971578243438 && echo $HOME/.ansible/tmp/ansible-tmp-1445471371.06-1971578243438'
<b-openshif-r18g8-prod.phx2.mydomain.net> EXEC sshpass -d14 ssh -C -tt -vvv -o ControlMaster=auto -o ControlPersist=60s -o ControlPath="/home/johntheodore/.ansible/cp/ansible-ssh-%h-%p-%r" -o GSSAPIAuthentication=no -o PubkeyAuthentication=no -o ConnectTimeout=10 b-openshif-r18g8-prod.phx2.mydomain.net /bin/sh -c 'mkdir -p $HOME/.ansible/tmp/ansible-tmp-1445471371.06-265515354516815 && chmod a+rx $HOME/.ansible/tmp/ansible-tmp-1445471371.06-265515354516815 && echo $HOME/.ansible/tmp/ansible-tmp-1445471371.06-265515354516815'
<b-openshif-r18f8-prod.phx2.mydomain.net> EXEC sshpass -d15 ssh -C -tt -vvv -o ControlMaster=auto -o ControlPersist=60s -o ControlPath="/home/johntheodore/.ansible/cp/ansible-ssh-%h-%p-%r" -o GSSAPIAuthentication=no -o PubkeyAuthentication=no -o ConnectTimeout=10 b-openshif-r18f8-prod.phx2.mydomain.net /bin/sh -c 'mkdir -p $HOME/.ansible/tmp/ansible-tmp-1445471371.06-251592570771308 && chmod a+rx $HOME/.ansible/tmp/ansible-tmp-1445471371.06-251592570771308 && echo $HOME/.ansible/tmp/ansible-tmp-1445471371.06-251592570771308'
<b-openshif-r18g8-prod.phx2.mydomain.net> PUT /tmp/tmp2dpsrF TO /home/johntheodore/.ansible/tmp/ansible-tmp-1445471371.06-265515354516815/openshift_facts
<b-openshif-r18f10-prod.phx2.mydomain.net> PUT /tmp/tmpeOR0ms TO /home/johntheodore/.ansible/tmp/ansible-tmp-1445471371.06-1971578243438/openshift_facts
<b-openshif-r18f8-prod.phx2.mydomain.net> PUT /tmp/tmphKk2KS TO /home/johntheodore/.ansible/tmp/ansible-tmp-1445471371.06-251592570771308/openshift_facts
<b-openshif-r18g8-prod.phx2.mydomain.net> EXEC sshpass -d14 ssh -C -tt -vvv -o ControlMaster=auto -o ControlPersist=60s -o ControlPath="/home/johntheodore/.ansible/cp/ansible-ssh-%h-%p-%r" -o GSSAPIAuthentication=no -o PubkeyAuthentication=no -o ConnectTimeout=10 b-openshif-r18g8-prod.phx2.mydomain.net /bin/sh -c 'sudo -k && sudo -H -S -p "[sudo via ansible, key=fsfxyfafzzzonvhgtdmffuxtbcaisbzi] password: " -u root /bin/sh -c '"'"'echo BECOME-SUCCESS-fsfxyfafzzzonvhgtdmffuxtbcaisbzi; LANG=C LC_CTYPE=C /usr/bin/python # pylint: disable=too-many-lines /home/johntheodore/.ansible/tmp/ansible-tmp-1445471371.06-265515354516815/openshift_facts; rm -rf /home/johntheodore/.ansible/tmp/ansible-tmp-1445471371.06-265515354516815/ >/dev/null 2>&1'"'"''
<b-openshif-r18f10-prod.phx2.mydomain.net> EXEC sshpass -d13 ssh -C -tt -vvv -o ControlMaster=auto -o ControlPersist=60s -o ControlPath="/home/johntheodore/.ansible/cp/ansible-ssh-%h-%p-%r" -o GSSAPIAuthentication=no -o PubkeyAuthentication=no -o ConnectTimeout=10 b-openshif-r18f10-prod.phx2.mydomain.net /bin/sh -c 'sudo -k && sudo -H -S -p "[sudo via ansible, key=wczrhbffjmkzsaseszlwlwninkhaulmz] password: " -u root /bin/sh -c '"'"'echo BECOME-SUCCESS-wczrhbffjmkzsaseszlwlwninkhaulmz; LANG=C LC_CTYPE=C /usr/bin/python # pylint: disable=too-many-lines /home/johntheodore/.ansible/tmp/ansible-tmp-1445471371.06-1971578243438/openshift_facts; rm -rf /home/johntheodore/.ansible/tmp/ansible-tmp-1445471371.06-1971578243438/ >/dev/null 2>&1'"'"''
<b-openshif-r18f8-prod.phx2.mydomain.net> EXEC sshpass -d15 ssh -C -tt -vvv -o ControlMaster=auto -o ControlPersist=60s -o ControlPath="/home/johntheodore/.ansible/cp/ansible-ssh-%h-%p-%r" -o GSSAPIAuthentication=no -o PubkeyAuthentication=no -o ConnectTimeout=10 b-openshif-r18f8-prod.phx2.mydomain.net /bin/sh -c 'sudo -k && sudo -H -S -p "[sudo via ansible, key=cqduineynsnmqmuaumzdkiyysuoieoce] password: " -u root /bin/sh -c '"'"'echo BECOME-SUCCESS-cqduineynsnmqmuaumzdkiyysuoieoce; LANG=C LC_CTYPE=C /usr/bin/python # pylint: disable=too-many-lines /home/johntheodore/.ansible/tmp/ansible-tmp-1445471371.06-251592570771308/openshift_facts; rm -rf /home/johntheodore/.ansible/tmp/ansible-tmp-1445471371.06-251592570771308/ >/dev/null 2>&1'"'"''

**Comment by abutcher** [comment]
@JohnTheodore Looks like #720 - I've cherry picked that into the PR for now and it's working for me so far.

**Comment by abutcher** [comment]
@JohnTheodore Also add these to your inventory.

```
openshift_router_selector='region=infra'
num_infra=0
```

**Comment by JohnTheodore** [comment]
@abutcher I added the two options you specified, and setup the master cluster variable as you described, the log is below: https://gist.github.com/JohnTheodore/f652d375769491eaf74c

**Comment by abutcher** [comment]
@JohnTheodore Looks like this isn't using my PR? Can you show me your inventory too

**Comment by JohnTheodore** [comment]
https://gist.github.com/JohnTheodore/59095b94e1a250922127

**Comment by abutcher** [comment]
To pull my changes down

```
git pull https://github.com/abutcher/openshift-ansible native-ha
```

and then you can reset when you're done or if you run into issues

```
git reset --hard origin/<branch>
```

The top commit is currentl 0024d7c

**Comment by JohnTheodore** [comment]
https://gist.github.com/JohnTheodore/cbb00d44857cfef084c5

TASK: [openshift_examples | Import RHEL streams] ****************************** 
<b-openshif-r18f10-prod.phx2.mydomain.net> ESTABLISH CONNECTION FOR USER: johntheodore
<b-openshif-r18f10-prod.phx2.mydomain.net> REMOTE_MODULE command oc create -n openshift -f /usr/share/openshift/examples/image-streams/image-streams-rhel7.json
<b-openshif-r18f10-prod.phx2.mydomain.net> EXEC sshpass -d4 ssh -C -tt -vvv -o ControlMaster=auto -o ControlPersist=60s -o ControlPath="/home/johntheodore/.ansible/cp/ansible-ssh-%h-%p-%r" -o GSSAPIAuthentication=no -o PubkeyAuthentication=no -o ConnectTimeout=10 b-openshif-r18f10-prod.phx2.mydomain.net /bin/sh -c 'mkdir -p $HOME/.ansible/tmp/ansible-tmp-1445555933.93-79965122410411 && chmod a+rx $HOME/.ansible/tmp/ansible-tmp-1445555933.93-79965122410411 && echo $HOME/.ansible/tmp/ansible-tmp-1445555933.93-79965122410411'
<b-openshif-r18f10-prod.phx2.mydomain.net> PUT /tmp/tmpieyo6x TO /home/johntheodore/.ansible/tmp/ansible-tmp-1445555933.93-79965122410411/command
<b-openshif-r18f10-prod.phx2.mydomain.net> EXEC sshpass -d4 ssh -C -tt -vvv -o ControlMaster=auto -o ControlPersist=60s -o ControlPath="/home/johntheodore/.ansible/cp/ansible-ssh-%h-%p-%r" -o GSSAPIAuthentication=no -o PubkeyAuthentication=no -o ConnectTimeout=10 b-openshif-r18f10-prod.phx2.mydomain.net /bin/sh -c 'sudo -k && sudo -H -S -p "[sudo via ansible, key=azulnssfnatjalrevmsjezjjubbbtbfj] password: " -u root /bin/sh -c '"'"'echo BECOME-SUCCESS-azulnssfnatjalrevmsjezjjubbbtbfj; LANG=C LC_CTYPE=C /usr/bin/python /home/johntheodore/.ansible/tmp/ansible-tmp-1445555933.93-79965122410411/command; rm -rf /home/johntheodore/.ansible/tmp/ansible-tmp-1445555933.93-79965122410411/ >/dev/null 2>&1'"'"''
fatal: [b-openshif-r18f10-prod.phx2.mydomain.net] => error while evaluating conditional: 'already exists' not in oex_import_rhel_streams.stderr and oex_import_rhel_streams.rc != 0

**Comment by abutcher** [comment]
@JohnTheodore Sorry there have been a few bugs in the last few days, that one looks like #728.

**Comment by JohnTheodore** [comment]
So the latest attempt seems to have worked!

**Comment by abutcher** [comment]
Excellent!

**Comment by abutcher** [comment]
@JohnTheodore I've added some configuration that allows sessions to be shared/decrypted by all masters. Would you mind running the latest `abutcher:native-ha` branch against your environment to try it out?

**Comment by JohnTheodore** [comment]
@abutcher sure I can try this out.

I had some other questions for you.

Can I put my ldap configuration into the yaml config somewhere so it sets up ldap and makes it work after the deploy? I have to modify the yaml config after the deployment.. though it would be nice if it's included. The variables are  my OU, a user/password to loo up users in ldap, the FQDN of the ldap box, etc.

Do you know which variable to modify to setup the proper domain (the one that shows up int he url bar) for the web server? so if I want https://openshift.phx2.mydomain.net, and not say https://b-openshif-r18f10-prod.phx2.mydomain.net:8443, so if I get routed to any of the 3 masters from a load balancer, I want it to show openshift.phx2.mydomain.net

and my last question.. how long do your deploys take to run? "real	299m27.276s" my `time` wrapper showed it took 5 hours to successfully complete my last ansible run of a 6 node cluster for the branch native-ha. These baremetals are beefy too, 128 gigs of ram, 32 cores, etc etc. I'm worried perhaps I'm hitting some timeouts. I don't have outbound Internet access, so I'm hosting all the RPMs for the install on my local repositories. If there is something I can do to make the runs go faster, that would be great. I can open a separate ticket for this too if you think it's warranted.

**Comment by abutcher** [comment]
@JohnTheodore You can put your ldap configuration in the inventory file, something like this...

```
openshift_master_identity_providers=[{'name': 'my_ldap_provider', 'challenge': 'true', 'login': 'true', 'kind': 'LDAPPasswordIdentityProvider', 'attributes': {'id': ['dn'], 'email': ['mail'], 'name': ['cn'], 'p\
referredUsername': ['uid']}, 'bindDN': '', 'bindPassword': '', 'ca': '/path/to/my-ldap-cacert.crt', 'insecure': 'false', 'url': 'ldap://ldap.mycompany.com:389/ou=users,dc=redhat,dc=com?uid'}]
```

Where `/path/to/my-ldap-cacert.crt` is already in place on the filesystem.

To get the naming right, you'll set `cluster_hostname` variables below and ensure that `openshift.phx2.mydomain.net` resolves to your load balancer address in DNS. You can terminate SSL at the load balancer if you have a custom certificate and listen on 443 instead of 8443. Some tasks such as `oc exec` will attempt to upgrade client connections to websocket using the spdy protocol. Your load balancer will either need to support this or you can load balance the masters with round robin DNS. The openshift users list might be a better place to ask these questions since there is a much wider group of people looking at it http://lists.openshift.redhat.com/openshiftmm/listinfo.

```
openshift_master_cluster_hostname=openshift.phx2.mydomain.net
openshift_master_cluster_public_hostname=openshift.phx2.mydomain.net
```

My ansible runs with the HA configuration are quite slow. We just merged in a PR that removes the 30 second waits so that should get you back a chunk of time. Are each of your 6 nodes masters as well? I think 3 masters should be sufficient.

**Comment by JohnTheodore** [comment]
yes, 3 masters, 3 nodes for etcd, and all 6 nodes are available as kubelets

**Comment by abutcher** [comment]
@JohnTheodore Just ran against my local vms (3 master, 3 etcd, 3 node, 1 lb) and it took 13m.

**Comment by JohnTheodore** [comment]
How can I tell what is consuming how much time? there is probably an ansible profiling tool. I wonder if vvvvv will output timestamps.

**Comment by abutcher** [comment]
@detiber might have an idea

**Comment by detiber** [comment]
We actually include ansible-profile in the openshift-ansible repo...  To use it you will need to do one of the following:
- add callback_plugins=<path to openshift-ansible checkout>/ansible-profile/callback_plugins
- set the ANSIBLE_CALLBACK_PLUGINS env variable to ansible-profile/callback_plugins before invoking ansible-playbook
At the end of the run you will see output similar to the following:
```
PLAY RECAP ******************************************************************** 
openshift_examples | Copy openshift examples ---------------------------- 4.17s
openshift_router | Deploy OpenShift Router ------------------------------ 3.08s
file name={{ g_etcd_mktemp.stdout }} state=absent ----------------------- 2.92s
nickhammond.logrotate | nickhammond.logrotate | Install logrotate ------- 1.20s
openshift_manage_node | Set node schedulability ------------------------- 0.99s
cockpit | Install cockpit-ws -------------------------------------------- 0.91s
openshift_node | Set node facts ----------------------------------------- 0.86s
openshift_facts  -------------------------------------------------------- 0.84s
openshift_manage_node | Wait for Node Registration ---------------------- 0.74s
openshift_node | Install Node package ----------------------------------- 0.70s
OSE3-master-ansible.test.example.com : ok=156  changed=7    unreachable=0    failed=0   
OSE3-node1-ansible.test.example.com : ok=45   changed=0    unreachable=0    failed=0   
OSE3-node2-ansible.test.example.com : ok=45   changed=0    unreachable=0    failed=0   
localhost                  : ok=10   changed=0    unreachable=0    failed=0   
```
To speed up runs in test environments, I use the following ansible.cfg (which I place in the root of the openshift-ansible checkout, however you can also use ~/.ansible.cfg or /etc/ansible/ansible.cfg if you adjust the paths):
```
[defaults]
callback_plugins = ansible-profile/callback_plugins
forks = 10
host_key_checking = False
remote_user = root
roles_path = roles/
hostfile = inventory/byo/hosts
gathering = smart
fact_caching = jsonfile
fact_caching_connection = /tmp/ansible/facts
fact_caching_timeout = 600

[privilege_escalation]
become = False

[ssh_connection]
ssh_args = -o ControlMaster=auto -o ControlPersist=600s
control_path = %(directory)s/%%h-%%r
pipelining = True
```

Notes on the config:
- Depending on the number of instances you are deploying, you may want to increase the number of forks even greater, but 10 will be sufficient for 6 hosts.  The default value is 5 which would limit the number of parallel tasks that ansible will perform is limited to 5, meaning that all node tasks will be limited to 5 operations at the same time only performing the operations on the last host when one has completed.
- If I'm not mistaken, the default ControlPersist value used by ansible is 60s, which means that each host will experience multiple reconnects delaying the entire run, so increasing the ControlPersist value to something greater than the expected runtime of the playbook will speed things up significantly as well.
- I'm setting host_key_checking to False to avoid contention around key checking, since openssh serializes access to the key file.
- Setting gathering to smart reduces the number of times that ansible reaches out to the hosts to query host facts, while the fact gathering isn't overly intensive it does save a few round trips, this can cause issues if ansible is being run from multiple different hosts though, since there can be periods of times that the different hosts see different facts.  You'll want the fact_caching_timeout to be greater than the run time for best effect.
- I set the control_path simply because I've had to deal with long hostnames that caused the default control_path to be too long causing errors
- Setting pipelining to true is probably the most impactful (besides setting the number of forks), since normally ansible will copy the module code for each and every task, with pipelining enabled it optimizes the transfer for multiple tasks.  The caveat here is that you will either need to use root on the remote hosts, or ensure that the sudo user being used is configured to not require a tty (requiretty being the default setting for sudo users on RHEL).

**Comment by JohnTheodore** [comment]
This ticket can be closed, if you want me to test more deployments I'm happy to do it. pm me as 'ayee' on openshift-dev. Thanks for the help, it's appreciated

**Comment by detiber** [comment]
@twiest please close
